compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -3684,6 +3700,17 @@ private STNode parseExpressionRhs(SyntaxKind tokenKind, OperatorPrecedence curre
             }
         }
 
+        // TODO: Add a good comment
+        if (tokenKind == SyntaxKind.LT_TOKEN && peek(2).kind == SyntaxKind.LT_TOKEN) {
+            tokenKind = SyntaxKind.DOUBLE_LT_TOKEN;
+        } else if (tokenKind == SyntaxKind.GT_TOKEN && peek(2).kind == SyntaxKind.GT_TOKEN) {
+            if (peek(3).kind == SyntaxKind.GT_TOKEN) {
+                tokenKind = SyntaxKind.TRIPPLE_GT_TOKEN;
+            } else {
+                tokenKind = SyntaxKind.DOUBLE_GT_TOKEN;
+            }
+        }
```

Comment:
In here we have to check whether next tokens belong to a `type-cast-expr`. It seems it is not possible unless we process up to the `>`. i.e. next token being an `@` or `isTypeStartingToken` does not always imply a `type-cast-expr` start.

Even if we processed up to `>`, still there's ambiguity. 
e.g.  `a <<b> c`  could mean either `a << (b > c)` or `a < (<b> c)`

So feel like it is better not  allow `a <<int> b;` to be parsed as a `type-cast-expr`.
Only  `a < <int> b;` is allowed. That way we don't need nodes for `<<`, `>>`, `>>>` at parser level. WDT?

Reply:
According to the spec, type-cast has higher-precedence over shift/compare expressions. So `a <<b> c` is correct to be parsed as `a <(<b> c)`. If the user wants the other behaviour, he can use parenthesis.

But yea, like you've said, we have to go and parse the component after `<` to determine the operator. Let me have a look and get back.